### DISCUSSION TITLE
Loading multiple MP4

### DISCUSSION THREAD

**Discussion by HighOnLemonz** [sipsorcery-org/sipsorcery#1252] [Q&A] [answered]
Hello,

Is there a way to load multiple MP4s and switch the live feed between them on demand?

Do I have to redo the entire example from MP4 example with the new MP4 while the previous is already running?

Thank you

**Comment by sipsorcery** [comment]
You could create multiple 'FFmpegFileSource' instances and then switch between them by setting and removing the peer connection event handler `mediaFileSource.OnVideoSourceEncodedSample += pc.SendVideo`.

It's not the most efficeint approach as you'll have the overhead of multiple sources in memory. Also all the video sources will need to be set to the same sending format.

**Reply by HighOnLemonz** [reply]
I just tried this approach and only one of the videos will stream.

__peerConnection = new RTCPeerConnection(config);
FFmpegInit.Initialise(FfmpegLogLevelEnum.AV_LOG_VERBOSE, null, _logger);

_mediaFileSource1 = new FFmpegFileSource("D:\\vids\\leftright.mp4", true, null);
_mediaFileSource1.RestrictFormats(x => x.Codec == VideoCodecsEnum.H264);
_videoTrack1 = new(_mediaFileSource1.GetVideoSourceFormats(), MediaStreamStatusEnum.SendOnly);
_peerConnection.addTrack(_videoTrack1);

_mediaFileSource2 = new FFmpegFileSource("D:\\vids\\updown.mp4", true, null);
_mediaFileSource2.RestrictFormats(x => x.Codec == VideoCodecsEnum.H264);
_videoTrack2 = new(_mediaFileSource2.GetVideoSourceFormats(), MediaStreamStatusEnum.SendOnly);
_peerConnection.addTrack(_videoTrack2);

_peerConnection.OnVideoFormatsNegotiated += (videoFormat) => _mediaFileSource1.SetVideoSourceFormat(videoFormat.First());
_peerConnection.OnVideoFormatsNegotiated += (videoFormat) => _mediaFileSource2.SetVideoSourceFormat(videoFormat.First());
_mediaFileSource1.OnVideoSourceEncodedSample += _peerConnection.SendVideo;_ // video that sends by default

At a later point I switch between the 2:
switch(_currentStream)
{
    case 0:
        _currentStream = 1;
        _mediaFileSource1.OnVideoSourceEncodedSample -= _peerConnection.SendVideo;
        _mediaFileSource2.OnVideoSourceEncodedSample += _peerConnection.SendVideo;
        break;
    case 1:
        _currentStream = 0;
        _mediaFileSource2.OnVideoSourceEncodedSample -= _peerConnection.SendVideo;
        _mediaFileSource1.OnVideoSourceEncodedSample += _peerConnection.SendVideo;
        break;
}

Case 0 will never work (nothing will be streamed), case 1 will always work. Switching between them works fine. I don't know what am I doing wrong. The videos have the same resolution and same codec. In the log I see nothing.

**Reply by sipsorcery** [reply, marked as answer]
Have a look at the Start and Stop methods on the FFmpegFileSource.

**Reply by HighOnLemonz** [reply]
All fine. Thank you.
